### ISSUE TITLE
Panic when trying to use /federate endpoint in v2.18.0-rc.0

### ISSUE THREAD

**Issue by csmarchbanks** [prometheus/prometheus#7181]
**What did you do?**
Tried to use `/federate` endpoint on the v2.18.0 release candidate

**What did you expect to see?**
Appropriate results coming back

**What did you see instead? Under which circumstances?**
Panic inside of Prometheus

**Environment**
* Prometheus version: v2.18.0-rc.0
* Prometheus configuration file: Reproducible using `documentation/examples/prometheus.yml`

**Stacktrace**
```
runtime error: invalid memory address or nil pointer dereference" stack="goroutine 388463 [running]:
github.com/prometheus/prometheus/web.withStackTracer.func1.1(0x2e92de0, 0xc000247e90, 0xc097834800)
    /app/web/web.go:98 +0xb5
panic(0x22b7e80, 0x41f4180)
    /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/prometheus/prometheus/tsdb.(*DB).Querier(0x0, 0x2ee3c20, 0xc09c2d3c80, 0x171c30cc6f1, 0x171c3115ad1, 0x0, 0x0, 0x0, 0x0)
    /app/tsdb/db.go:1283 +0x5d
github.com/prometheus/prometheus/web.(*Handler).federation(0xc0008b2b40, 0x2ece520, 0xc09c2a9680, 0xc097834b00)
    /app/web/federate.go:78 +0x758
net/http.HandlerFunc.ServeHTTP(0xc0002054f0, 0x2ece520, 0xc09c2a9680, 0xc097834b00)
    /usr/local/go/src/net/http/server.go:2012 +0x44
github.com/prometheus/prometheus/util/httputil.CompressionHandler.ServeHTTP(0x2e97c00, 0xc0002054f0, 0x7f570c507a58, 0xc09be56cd0, 0xc097834b00)
    /app/util/httputil/compression.go:90 +0x7c
github.com/prometheus/prometheus/web.(*Handler).testReady.func1(0x7f570c507a58, 0xc09be56cd0, 0xc097834b00)
    /app/web/web.go:495 +0x55
net/http.HandlerFunc.ServeHTTP(0xc000209600, 0x7f570c507a58, 0xc09be56cd0, 0xc097834b00)
    /usr/local/go/src/net/http/server.go:2012 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerResponseSize.func1(0x7f570c507a58, 0xc09be56c80, 0xc097834b00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:196 +0xea
net/http.HandlerFunc.ServeHTTP(0xc00073ff20, 0x7f570c507a58, 0xc09be56c80, 0xc097834b00)
    /usr/local/go/src/net/http/server.go:2012 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2(0x7f570c507a58, 0xc09be56c80, 0xc097834b00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:76 +0xb2
net/http.HandlerFunc.ServeHTTP(0xc0008f4030, 0x7f570c507a58, 0xc09be56c80, 0xc097834b00)
    /usr/local/go/src/net/http/server.go:2012 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1(0x2edc360, 0xc09be56c30, 0xc097834b00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:100 +0xda
github.com/prometheus/prometheus/web.setPathWithPrefix.func1.1(0x2edc360, 0xc09be56c30, 0xc097834a00)
    /app/web/web.go:1128 +0x1c6
github.com/prometheus/common/route.(*Router).handle.func1(0x2edc360, 0xc09be56c30, 0xc097834900, 0x0, 0x0, 0x0)
    /app/vendor/github.com/prometheus/common/route/route.go:83 +0x27f
github.com/julienschmidt/httprouter.(*Router).ServeHTTP(0xc000201e60, 0x2edc360, 0xc09be56c30, 0xc097834900)
    /app/vendor/github.com/julienschmidt/httprouter/router.go:387 +0xc37
github.com/prometheus/common/route.(*Router).ServeHTTP(0xc00000d2e0, 0x2edc360, 0xc09be56c30, 0xc097834900)
    /app/vendor/github.com/prometheus/common/route/route.go:121 +0x4c
net/http.(*ServeMux).ServeHTTP(0xc0000d5240, 0x2edc360, 0xc09be56c30, 0xc097834900)
    /usr/local/go/src/net/http/server.go:2387 +0x1a5
github.com/opentracing-contrib/go-stdlib/nethttp.MiddlewareFunc.func5(0x2eda920, 0xc0861cb6c0, 0xc097834800)
    /app/vendor/github.com/opentracing-contrib/go-stdlib/nethttp/server.go:140 +0x5f6
net/http.HandlerFunc.ServeHTTP(0xc000537d40, 0x2eda920, 0xc0861cb6c0, 0xc097834800)
    /usr/local/go/src/net/http/server.go:2012 +0x44
github.com/prometheus/prometheus/web.withStackTracer.func1(0x2eda920, 0xc0861cb6c0, 0xc097834800)
    /app/web/web.go:103 +0x8b
net/http.HandlerFunc.ServeHTTP(0xc000537d70, 0x2eda920, 0xc0861cb6c0, 0xc097834800)
    /usr/local/go/src/net/http/server.go:2012 +0x44
net/http.serverHandler.ServeHTTP(0xc0005be000, 0x2eda920, 0xc0861cb6c0, 0xc097834800)
    /usr/local/go/src/net/http/server.go:2807 +0xa3
net/http.(*conn).serve(0xc09be17720, 0x2ee3b60, 0xc0835db840)
    /usr/local/go/src/net/http/server.go:1895 +0x86c
created by net/http.(*Server).Serve
    /usr/local/go/src/net/http/server.go:2933 +0x35c
```

cc: @bwplotka as this is blocking for the release.

**Comment by bwplotka** [comment]
Ack, will check tomorrow

**Comment by bwplotka** [comment]
```
// Querier returns a new querier over the data partition for the given time range.
// A goroutine must not handle more than one open Querier.
func (db *DB) Querier(_ context.Context, mint, maxt int64) (storage.Querier, error) {
	var blocks []BlockReader
	var blockMetas []BlockMeta

	db.mtx.RLock()  <-- panic 
	defer db.mtx.RUnlock()

```
:thinking:

**Comment by bwplotka** [comment]
Long term fix ready for review: https://github.com/prometheus/prometheus/pull/7182

**Comment by bwplotka** [comment]
After that will release rc.1

**Comment by bwplotka** [comment]
Fixed and will be released in 2.18.0-rc.1

**Comment by bwplotka** [comment]
Thanks for super quick feedback! :hugs:
